### ISSUE TITLE
[RFC] Deprecate `log_every_n_steps` from Trainer 

### ISSUE THREAD

**Issue by daniellepintz** [PyTorchLightning/pytorch-lightning#9726]
## Proposed refactoring or deprecation

Deprecate `log_every_n_steps` from Trainer and make it available as a parameter to loggers that have this capability.

### Motivation

Same reasons as #8991. How often to log should be considered an internal implementation detail of each logger

### Pitch

Deprecate `log_every_n_steps` from Trainer, and move it to the init for logger classes that support this functionality (e.g. CSVLogger).

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

cc @justusschock @awaelchli @akihironitta @rohitgr7 @carmocca @edward-io @ananthsub @kamil-kaczmarek @Raalsky @Blaizzy

**Comment by tchaton** [comment]
Sounds good to me !

@edenlightning ?

**Comment by awaelchli** [comment]
It's not an internal implementation detail of a logger end point. It applies to the Trainer and its logging infrastructure because this setting determines how often logs are sent to the logger. I vote against removing this argument.

**Comment by carmocca** [comment]
Pros for removing it:

- This could be entirely done in the training step (by gating the `log` call with an `if` statement) depending on what you are logging. Wouldn't work for aggregated tensors but it would for `Metric`s.
- Increased flexibility - the current implementation is hidden in the `LoggerConnector`.

Pros for keeping it:
- Harder configuration from external tools.

Alternatives:

- Add a `log_every_n_steps` flag to `LightningModule.log`. This would remove it from the `Trainer` and improve the flexibility. Especially for heterogeneous data types where you might want different values per key.

We haven't reached a consensus on this however:

> internal implementation detail of each logger

We don't think it should be the responsibility of the loggers, but either the `LoggerConnector` or `self.log` method.

**Comment by tchaton** [comment]
Based on @carmocca pros/cons. 

I believe removing this parameter would violate one of our core principle: `we make our API discoverable incrementally` and this parameter is quite key to easily parameterized logging from external tooling.

Best,
T.C

**Comment by ananthsub** [comment]
The huge issue today with `log_every_n_steps` is that it is broken for Trainer.validate, Trainer.test, and Trainer.predict with high probability.

`log_every_n_steps` works with the trainer's `global_step` to determine if data is going to be logged: https://github.com/PyTorchLightning/pytorch-lightning/blob/f9b9cdb0d1d4e26d25fe13f19f12ea88690aa0a8/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L74-L77
 
`global_step` is defined as the number of parameter updates that occur. This number is incremented *only* during fitting. It is *never* incremented for validation, test, or prediction routines. 

Therefore, if your LightningModule calls `self.log` inside of validate, test, or predict steps, it is highly unlikely that your data will be logged! You would have to get lucky that the global step is a multiple of `log_every_n_steps` from a prior fitting routine.

The workaround is to always set `Trainer(log_every_n_steps=1)` and then gate self.log calls in your LightningModule with your own flag for the log frequency. This way, you can at least control the granularity with which you want to log without any surprising behavior from the Lightning framework interfering with this.

What could the framework do?
- Either remove the flag and make users gate this themselves
- Or update the logger connector code to to take into account the batch_idx if validate, test, or predict were called instead of the global step

**Comment by carmocca** [comment]
Even if we choose to remove the flag, we will need to fix the logger connector

**Comment by ananthsub** [comment]
> Even if we choose to remove the flag, we will need to fix the logger connector

If we removed the flag, wouldn't this code also disappear from the logger connector? `log` would always `log` then

**Comment by carmocca** [comment]
Yes, but it's a long time between deprecation and removal, we would want to fix it

**Comment by awaelchli** [comment]
@ananthsub I disagree with your assessment.

1. It not submitting logs on step during the evaluation does not mean it is "broken". It is in fact intentional. It has nothing to do with the log interval. 

2. The `log_every_n_steps` does not turn off `self.log`, it only prevents logs being sent to the logger object. That means the on_epoch accumulated and reduced metrics will always be correct, in the sense that `log_every_n_steps` does cause these metrics to be dropped. 

Therefore the proposal claiming that the user may guard their `self.log` calls in the LightningModule with log interval logic is not an acceptable alternative imo. The user CAN do that but they shouldn't expect their epoch metrics to be correct. For the same reason I disagree with the two points in @carmocca's https://github.com/PyTorchLightning/pytorch-lightning/issues/9726#issuecomment-934241746 listed as pros for removing it. 

I'm taking the defensive position here in the hope that this perfectly fine functionality will not be just removed for the wrong reasons.

The alternative that @carmocca suggested would preserve the functionality:

> Add a log_every_n_steps flag to LightningModule.log. This would remove it from the Trainer and improve the flexibility. Especially for heterogeneous data types where you might want different values per key.


3. The misunderstanding that `self.log` maps 1:1 to the logger object is the fault of our documentation not explaining it well.

**Comment by ananthsub** [comment]
@awaelchli IMO LightningModule.log does way too many things. If you had to describe it in one line, how would you do so? 

That's where these misunderstandings emerge from. It's also named log, and Lightning offers Loggers, so it's understandable to think that log would always pass data to loggers.

**Comment by carmocca** [comment]
> It's also named log, and Lightning offers Loggers

This was an unfortunate choice, log* as a term is way too overloaded, logging and loggers are separate concepts where the `LoggerConnector` manages both and publishes to the loggers the data gathered from logging
